angular/projects/global/src/lib/services/store.service.ts

```diff
@@ -0,0 +1,153 @@
+  public getLastNavigationObject() {
+    const stack = this.getNavigationStack();
+    if (stack) {
+      return stack[stack.length - 1];
+    } else {
+      return null;
+    }
+  }
+
+  public pushNavigationStack(backUrl: string[], breadcrumbs: object[], componentId: string = null) {
+    const navigationObject = {
+      backUrl: backUrl,
+      breadcrumbs: breadcrumbs,
+      componentId: componentId
+    };
+    const stack = this.getNavigationStack();
+    if (stack) {
+      stack.push(navigationObject);
+      this.state.navigationStack = stack;
+    } else {
+      this.state.navigationStack = [navigationObject];
+    }
+  }
+
+  public popNavigationStack() {
+    const stack = this.getNavigationStack();
+    if (stack) {
+      let stackObject = stack.pop();
+      if (stack.length === 0) {
+        this.clearNavigationStack();
+        stackObject = null;
+      } else {
+        this.state.navigationStack = stack;
+      }
+      return stackObject;
+    } else {
+      this.clearNavigationStack();
+      return null;
+    }
+  }
+
+  public navigateBreadcrumb(breadcrumb, router) {
+    const stack = this.getNavigationStack();
+    let poppedItem = null;
+    let isPopping = true;
+    if (stack) {
+      while (isPopping) {
+        poppedItem = this.popNavigationStack();
+        if (poppedItem == null) {
+          break;
+        } else if (poppedItem.breadcrumbs[poppedItem.breadcrumbs.length - 1] === breadcrumb) {
+          isPopping = false;
+        }
+      }
+      router.navigate(breadcrumb.route);
+    } else {
+      router.navigate(['/']);
+    }
+  }
+
+  public clearNavigationStack() {
+    this.state.navigationStack = null;
+  }
+}
+
+export interface INavigationObjectParams {
+  breadcrumbs: object[];
+  componentName?: string;
+  componentId?: string;
+  pageData?: any;
+}
+
+/*
+  Example of a back Url array:
+  ['/p', this.project._id, 'edit']
+
+  Example of a breadcrumbs array:
+  Note that the objects in the array contain a route and a label.
+  [
+      {
+          route: ['/projects'],
+          label: 'All Projects'
+      },
+      {
+          route: ['/p', this.project._id],
+          label: this.project.name
+      },
+      {
+          route: ['/p', this.project._id, 'edit'],
+          label: 'Edit'
+      }
+  ]
+*/
+export class NavigationObject {
+  public backRef;
+  public breadcrumbs;
+  public componentName;
+  public componentId;
+  public pageData;
+
+  constructor(params: INavigationObjectParams) {
+    if (!params) {
+      throw Error('Params are required');
+    }
+
+    if (!params.breadcrumbs) {
+      throw Error('You must pass in a breadcrumbs');
+    }
+
+    this.breadcrumbs = (params && params.breadcrumbs) || [];
+    this.componentName = (params && params.componentName) || '';
+    this.componentId = (params && params.componentId) || '';
+
+    // Set backRef
+    if (this.breadcrumbs.length > 1) {
+      this.backRef = this.breadcrumbs[this.breadcrumbs.length - 2].route;
```

Comment:
Probably I just haven't wrapped my head around this enough, but wouldn't the backref just be 1 element back, and not 2?

Reply:
Nevermind I get it.  Because `breadcrumbs.length` is 1 bigger than the last index, so -2 is just the previous index.